src/bnet_client.py

```diff
@@ -365,67 +374,47 @@ async def disconnect(self):
         log.info("disconnected from battle.net")
         if self.writer:
             self.writer.close()
+        self.writer = None
+        self.reader = None
         self.authenticated = False
 
-    async def fetch_friend_battle_tag(self, entity_id, future: asyncio.Future):
+    async def fetch_friend_battle_tag(self, entity_id, future: 'asyncio.Future[AccountInfo]'):
         request = presence_service_pb2.QueryRequest()
         request.entity_id.high = entity_id.high
         request.entity_id.low = entity_id.low
-
-        key = request.key.add()
-        key.program = 0x424e
-        key.group = 1  # account
-        key.field = 4  # battle_tag
+        for i in [4]:  # for possible values see self.AccountInfo
+            key = request.key.add()
+            key.program = 0x424e  # hex code for "BN"
+            key.group = 1  # account
+            key.field = i
 
         await self._send_message(self._PRESENCE_SERVICE, 4, request, functools.partial(self._on_presence__query_account, future))
 
-    async def fetch_friend_presence_account_details(self, account_id, future: asyncio.Future):
+    async def fetch_friend_presence_account_details(self, entity_id, future: 'asyncio.Future[AccountInfo]'):
```

Comment:
we can avoid repeating code for `fetch_friend_battle_tag` and `fetch_friend_presence_account_details`

Reply:
What is your suggestion @bartok765 ? Maybe just put this in a private function with variable values replaced?
```
        request = presence_service_pb2.QueryRequest()
        request.entity_id.high = entity_id.high
        request.entity_id.low = entity_id.low
        for i in [...]:
            key = request.key.add()
            key.program = 0x424e  # hex code for "BN"
            key.group = ...
            key.field = i
```

Or do you have another idea?

Reply:
`group` is the same in both methods, so yes - it could be either one private method for basic request with additional parameter of field list to set OR both methods may be just merged into one with additional param. I have no opinion which is better.